<a href="https://colab.research.google.com/github/Madihasafi/ITC-300-Data-Visualization-/blob/main/Madiha_Safi%2C_Template_of_w4p1_Exploratory_Data_Analysis_(Part_2).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# ITC 300 – Data Visualization
## w4p1 Exploratory Data Analysis (Part - 2)

**Laboratory focus:** DataFrame inspection, descriptive statistics, column selection, categorical exploration, `groupby()`, data types, Boolean filtering, and aggregation.

This notebook is designed for **Google Colab** and follows the topics and worked examples from the **Week 2 and 3 EDA using pandas – Part 02** lecture material.

### Dataset
The laboratory uses **`Salaries.csv`**, containing **397 records and 7 columns**:

- `Unnamed: 0`
- `rank`
- `discipline`
- `yrs.since.phd`
- `yrs.service`
- `sex`
- `salary`

> **Important:** Upload `Salaries.csv` when requested in the setup section.

## Learning Outcomes

By the end of this laboratory, you should be able to:

1. Import the pandas library and load a CSV file into a DataFrame.
2. Inspect DataFrame structure using `head()`, `shape`, `columns`, `dtypes`, and `values`.
3. Distinguish between pandas **attributes** and **methods**.
4. Use `describe()`, `mean()`, `min()`, `max()`, `median()`, and `std()` to summarize numeric data.
5. Select one column as a Series and multiple columns as a DataFrame.
6. Identify unique categorical values with `unique()` and category frequencies with `value_counts()`.
7. Group records using `groupby()` and compute aggregate statistics for each group.
8. Perform multi-key grouping and interpret a hierarchical grouped result.
9. Construct Boolean filters using multiple conditions and aggregate the filtered subsets.
10. Recognize and correct common pandas errors, including import typos, wrong column labels, and incorrect `groupby()` syntax.
11. Interpret grouped and filtered results as **descriptive summaries** rather than causal explanations.

## Laboratory Outline

### Part A — Environment and Data Loading
1. Import pandas
2. Upload and read `Salaries.csv`
3. Verify successful loading

### Part B — Basic DataFrame Inspection
4. Preview records with `head()`
5. Inspect `shape`, `columns`, `dtypes`, and `values`
6. Understand attributes versus methods

### Part C — Descriptive Statistics
7. Use `describe()`
8. Compute salary and service summary statistics
9. Select multiple columns before summarizing them

### Part D — Column and Category Exploration
10. Select a single column as a Series
11. Preserve a one-column DataFrame with double brackets
12. Inspect unique values
13. Count categorical values

### Part E — GroupBy Analysis
14. Group by academic rank
15. Compare salary by sex
16. Compare salary by rank
17. Compare service by sex
18. Group by multiple categorical keys
19. Understand common `groupby()` mistakes

### Part F — Boolean Filtering and Aggregation
20. Filter professors with fewer than 10 years of service
21. Filter professors with more than 10 years of service
22. Compare subgroup counts and mean salaries

### Part H — Consolidated EDA Workflow
23. Consolidated EDA workflow

### Part I — Guided Laboratory Exercises
24. Guided exercises

# Part A — Environment and Data Loading

## A1. Import pandas

`pandas` is the main Python library used in this laboratory for tabular data analysis.

The conventional alias is:

```python
import pandas as pd
```

Using the alias `pd` keeps later commands concise, for example `pd.read_csv(...)`.

In [1]:
import pandas as pd
import numpy as np

print("pandas version:", pd.__version__)
print("numpy version:", np.__version__)

pandas version: 2.2.3
numpy version: 2.1.3


## A2. Upload `Salaries.csv` in Google Colab

A relative filename such as `"Salaries.csv"` works when the CSV file is in the notebook's current working directory.

In Google Colab, upload the file using the following cell.

In [3]:
from google.colab import files

uploaded = files.upload()

Saving Salaries.csv to Salaries (1).csv


## A3. Load the CSV file into a DataFrame

`pd.read_csv()` reads the external CSV file and returns a pandas **DataFrame**.

The DataFrame is stored in the variable `df`.

In [15]:
#Import Salaries.csv file
df=pd.read_csv("Salaries.csv")

### Expected dataset structure

The lecture dataset contains **397 rows × 7 columns**.

The first five records are expected to begin approximately as follows:

| Index | Unnamed: 0 | rank | discipline | yrs.since.phd | yrs.service | sex | salary |
|---:|---:|---|---|---:|---:|---|---:|
| 0 | 1 | Prof | B | 19 | 18 | Male | 139750 |
| 1 | 2 | Prof | B | 20 | 16 | Male | 173200 |
| 2 | 3 | AsstProf | B | 4 | 3 | Male | 79750 |
| 3 | 4 | Prof | B | 45 | 39 | Male | 115000 |
| 4 | 5 | Prof | B | 40 | 41 | Male | 141500 |

The column `Unnamed: 0` is an index-like field stored in the CSV and is imported as a normal column.

# Part B — Basic DataFrame Inspection

## B1. Preview records with `head()`

`head()` is a **method** that returns the first rows of a DataFrame.

- `df.head()` returns the first 5 rows.
- `df.head(10)` returns the first 10 rows.

This is preferable to displaying a very large DataFrame when performing an initial inspection.

In [17]:
#Display the first five records
df.head()

,Unnamed: 0,rank,discipline,yrs.since.phd,yrs.service,sex,salary
0,1,Prof,B,19,18,Male,139750
1,2,Prof,B,20,16,Male,173200
2,3,AsstProf,B,4,3,Male,79750
3,4,Prof,B,45,39,Male,115000
4,5,Prof,B,40,41,Male,141500


In [18]:
#Display the first ten records
df.head(10)

,Unnamed: 0,rank,discipline,yrs.since.phd,yrs.service,sex,salary
0,1,Prof,B,19,18,Male,139750
1,2,Prof,B,20,16,Male,173200
2,3,AsstProf,B,4,3,Male,79750
3,4,Prof,B,45,39,Male,115000
4,5,Prof,B,40,41,Male,141500
5,6,AssocProf,B,6,6,Male,97000
6,7,Prof,B,30,23,Male,175000
7,8,Prof,B,45,45,Male,147765
8,9,Prof,B,21,20,Male,119250
9,10,Prof,B,18,18,Female,129000


In [19]:
#Display the last five records
df.tail()

,Unnamed: 0,rank,discipline,yrs.since.phd,yrs.service,sex,salary
392,393,Prof,A,33,30,Male,103106
393,394,Prof,A,31,19,Male,150564
394,395,Prof,A,42,25,Male,101738
395,396,Prof,A,25,15,Male,95329
396,397,AsstProf,A,8,4,Male,81035


## B2. Check DataFrame dimensions with `shape`

`shape` is an **attribute**, not a method. It returns a tuple:

```text
(number_of_rows, number_of_columns)
```

For this dataset, the expected result is:

```text
(397, 7)
```

In [20]:
#Display dataframe dimensions
df.shape

(397, 7)

## B3. Inspect exact column labels with `columns`

Column labels must be written exactly as they appear in the DataFrame.

This is especially important here because names such as:

- `yrs.since.phd`
- `yrs.service`

contain dots.

In [21]:
#display column names
df.columns

Index(['Unnamed: 0', 'rank', 'discipline', 'yrs.since.phd', 'yrs.service',
       'sex', 'salary'],
      dtype='object')

In [22]:
#display row labels
df.index

RangeIndex(start=0, stop=397, step=1)

## B4. Inspect data types with `dtypes`

`dtypes` reports the data type of every DataFrame column.

In the lecture dataset:

- text/categorical columns are stored as `object`
- integer numeric columns are stored as `int64`

In [23]:
#Inspect all attributes data types
df.dtypes

,0
Unnamed: 0,int64
rank,object
discipline,object
yrs.since.phd,int64
yrs.service,int64
sex,object
salary,int64


Expected data types:

| Column | Expected dtype |
|---|---|
| `Unnamed: 0` | `int64` |
| `rank` | `object` |
| `discipline` | `object` |
| `yrs.since.phd` | `int64` |
| `yrs.service` | `int64` |
| `sex` | `object` |
| `salary` | `int64` |

## B5. Check the dtype of one selected column

Selecting one column produces a pandas **Series**.

The `.dtype` attribute reports the type of that Series.

In [24]:
#Inspect salary attribute data type
df["salary"].dtype

dtype('int64')

Expected result:

```text
dtype('int64')
```

This confirms that `salary` is numeric and can be used directly in arithmetic and aggregate calculations.

## B6. Inspect the underlying values with `values`

`df.values` exposes the underlying values in a NumPy-style array.

Because the DataFrame contains both text and numeric data, the complete matrix is typically represented with `dtype=object`.

Important distinction:

- A **DataFrame** preserves row/column labels.
- `df.values` exposes the raw value matrix and does not preserve the convenient labeled presentation.

In [25]:
#Inspect dataframe values
df.values

array([[1, 'Prof', 'B', ..., 18, 'Male', 139750],
       [2, 'Prof', 'B', ..., 16, 'Male', 173200],
       [3, 'AsstProf', 'B', ..., 3, 'Male', 79750],
       ...,
       [395, 'Prof', 'A', ..., 25, 'Male', 101738],
       [396, 'Prof', 'A', ..., 15, 'Male', 95329],
       [397, 'AsstProf', 'A', ..., 4, 'Male', 81035]], dtype=object)

## B7. Attributes versus methods

A useful pandas syntax distinction is:

### Attributes — no parentheses
Examples:

- `df.dtypes`
- `df.columns`
- `df.axes`
- `df.ndim`
- `df.size`
- `df.shape`
- `df.values`

### Methods — use parentheses
Examples:

- `df.head()`
- `df.tail()`
- `df.describe()`
- `df.max()`
- `df.min()`
- `df.mean()`
- `df.median()`
- `df.std()`
- `df.sample()`
- `df.dropna()`

For example:

```python
df.shape       # correct attribute usage
df.head()      # correct method usage
```

In [26]:
#Fill the print functions
print("axes:",df.axes )
print("ndim:", df.ndim)
print("size:", df.size)
print("shape:", df.shape)

axes: [RangeIndex(start=0, stop=397, step=1), Index(['Unnamed: 0', 'rank', 'discipline', 'yrs.since.phd', 'yrs.service',
       'sex', 'salary'],
      dtype='object')]
ndim: 2
size: 2779
shape: (397, 7)


# Part C — Descriptive Statistics

## C1. Summarize numeric columns with `describe()`

`describe()` returns standard descriptive statistics for numeric columns:

- `count`
- `mean`
- `std`
- `min`
- `25%`
- `50%` (median)
- `75%`
- `max`

For the lecture dataset, the salary summary includes approximately:

- mean = **113706.46**
- minimum = **57800**
- median = **107300**
- maximum = **231545**

In [53]:
import builtins

#Display all attributes summary statistics
df.columns = ["Unnamed: 0", "rank", "discipline",
              "yrs.since.phd", "yrs.service", "sex", "salary"]

# Restore the built-in print function if it was overwritten
print = builtins.print

# Display summary statistics for ALL attributes
print(df.describe(include='all').to_string())

        Unnamed: 0  rank discipline  yrs.since.phd  yrs.service   sex         salary
count   397.000000   397        397     397.000000   397.000000   397     397.000000
unique         NaN     3          2            NaN          NaN     2            NaN
top            NaN  Prof          B            NaN          NaN  Male            NaN
freq           NaN   266        216            NaN          NaN   358            NaN
mean    199.000000   NaN        NaN      22.314861    17.614610   NaN  113706.458438
std     114.748275   NaN        NaN      12.887003    13.006024   NaN   30289.038695
min       1.000000   NaN        NaN       1.000000     0.000000   NaN   57800.000000
25%     100.000000   NaN        NaN      12.000000     7.000000   NaN   91000.000000
50%     199.000000   NaN        NaN      21.000000    16.000000   NaN  107300.000000
75%     298.000000   NaN        NaN      32.000000    27.000000   NaN  134185.000000
max     397.000000   NaN        NaN      56.000000    60.000000  

## C2. Mean salary

A direct single-column calculation is the clearest way to answer:

> What is the average salary?

In [58]:
mean_salary = df["salary"].mean()
mean_salary

np.float64(113706.45843828715)

Expected result:

```text
113706.45843828715
```

## C3. Maximum salary and maximum years of service

In [60]:
#Find maximum salary and years of service

max_salary = df["salary"].max()
max_service = df["yrs.service"].max()
print("Maximum salary:", max_salary)
print("Maximum years of service:", max_service)


Maximum salary: 231545
Maximum years of service: 60


Expected results:

```text
Maximum salary: 231545
Maximum years of service: 60
```

## C4. Additional summary measures

The lecture materials list `min()`, `mean()`, `median()`, and `std()` as useful summary methods.

The next cell collects several salary statistics together.

In [61]:
salary_summary = {
    "minimum": df["salary"].min(),
    "maximum": df["salary"].max(),
    "mean": df["salary"].mean(),
    "median": df["salary"].median(),
    "standard_deviation": df["salary"].std()
}

salary_summary

{'minimum': 57800,
 'maximum': 231545,
 'mean': np.float64(113706.45843828715),
 'median': 107300.0,
 'standard_deviation': 30289.038694747804}

## C5. Select multiple columns and then summarize

Double brackets select multiple columns and preserve a **DataFrame** result.

Example:

```python
df[["yrs.service", "salary"]]
```

In [74]:
df[["yrs.service", "salary"]]

,yrs.service,salary
0,18,139750
1,16,173200
2,3,79750
3,39,115000
4,41,141500
...,...,...
392,30,103106
393,19,150564
394,25,101738
395,15,95329


Now apply `min()` to only the selected variables.

In [72]:
df[["yrs.service", "salary"]].min()

,0
yrs.service,0
salary,57800


Expected result:

```text
yrs.service        0
salary         57800
```

In [73]:
df[["yrs.service", "salary", "yrs.since.phd"]].min()

,0
yrs.service,0
salary,57800
yrs.since.phd,1


Expected result:

```text
yrs.service        0
salary         57800
yrs.since.phd      1
```

# Part D — Column and Category Exploration

## D1. Select one column as a Series

Bracket notation is the most reliable general form:

```python
df["rank"]
```

Selecting one column normally returns a **Series**.

In [75]:
rank_series = df["rank"]
rank_series

,rank
0,Prof
1,Prof
2,AsstProf
3,Prof
4,Prof
...,...
392,Prof
393,Prof
394,Prof
395,Prof


Use bracket notation because `rank` can conflict with existing DataFrame functionality.

Therefore:

```python
df["rank"]
```

is preferred to:

```python
df.rank
```

## D2. Preserve a one-column DataFrame

Double brackets keep the result two-dimensional.

In [77]:
rank_dataframe = df[["rank"]]
rank_dataframe.head()

,rank
0,Prof
1,Prof
2,AsstProf
3,Prof
4,Prof


Compare the object types.

In [78]:
print(type(rank_series))
print(type(rank_dataframe))

<class 'pandas.core.series.Series'>
<class 'pandas.core.frame.DataFrame'>


## D3. Find unique rank categories

`unique()` returns the distinct values stored in the selected Series.

In [79]:
ranks= df["rank"].unique()
print(ranks)
print(type(ranks))

['Prof' 'AsstProf' 'AssocProf']
<class 'numpy.ndarray'>


Expected categories:

```text
['Prof', 'AsstProf', 'AssocProf']
```

This is useful for discovering category labels and detecting spelling/coding inconsistencies.

## D4. Count category frequencies with `value_counts()`

`value_counts()` reports how many rows belong to each category.

In [80]:
rank_counts= df["rank"].value_counts()
print(rank_counts)
print(type(rank_counts))


rank
Prof         266
AsstProf      67
AssocProf     64
Name: count, dtype: int64
<class 'pandas.core.series.Series'>


Expected counts:

| rank | count |
|---|---:|
| Prof | 266 |
| AsstProf | 67 |
| AssocProf | 64 |

The counts total **397**, matching the number of DataFrame rows.

In [81]:
rank_counts = df["rank"].value_counts()

print("Total counted records:", rank_counts.sum())
print("Total counted records type:", type(rank_counts.sum()))
print("DataFrame rows:", df.shape[0])
print("Counts match row count:", rank_counts.sum() == df.shape[0])
print("Counts match row count type:", type(rank_counts.sum() == df.shape[0]))

Total counted records: 397
Total counted records type: <class 'numpy.int64'>
DataFrame rows: 397
Counts match row count: True
Counts match row count type: <class 'numpy.bool'>


# Part E — GroupBy Analysis

## E1. Conceptual model of `groupby()`

`groupby()` answers two connected questions:

1. **How should the rows be divided?**  
   This is determined by the grouping column(s).

2. **What should be calculated for each group?**  
   This is determined by the selected target variable and aggregation method.

A reusable pattern is:

```python
df.groupby(GROUPING_COLUMN)[TARGET_COLUMN].AGGREGATION()
```

For example:

```python
df.groupby("sex")["salary"].mean()
```

## E2. Show the first row from each academic-rank group

`first()` returns the first record in each group.

In [82]:
#display first record in each group
df.groupby("rank").first()

,Unnamed: 0,discipline,yrs.since.phd,yrs.service,sex,salary
rank,,,,,,
AssocProf,6,B,6,6,Male,97000
AsstProf,3,B,4,3,Male,79750
Prof,1,B,19,18,Male,139750


Expected representative rows begin with:

| rank | Unnamed: 0 | discipline | yrs.since.phd | yrs.service | sex | salary |
|---|---:|---|---:|---:|---|---:|
| AssocProf | 6 | B | 6 | 6 | Male | 97000 |
| AsstProf | 3 | B | 4 | 3 | Male | 79750 |
| Prof | 1 | B | 19 | 18 | Male | 139750 |

## E3. Mean values by academic rank

`numeric_only=True` limits the aggregation to numeric columns.

In [83]:
#Display groupped by rank mean values of numerical attributes
df.groupby("rank").mean(numeric_only=True)

,Unnamed: 0,yrs.since.phd,yrs.service,salary
rank,,,,
AssocProf,179.828125,15.453125,11.953125,93876.437500
AsstProf,169.761194,5.104478,2.373134,80775.985075
Prof,210.977444,28.300752,22.815789,126772.109023


Expected mean salary by rank:

- **AssocProf:** approximately 93,876.44
- **AsstProf:** approximately 80,775.99
- **Prof:** approximately 126,772.11

## E4. Practical Question 1 — Mean salary by sex

Analytical question:

> What is the average salary for female and male faculty records?

In [85]:
print(df.columns)
male_female_salaries = df.groupby("sex")["salary"].mean()
print(male_female_salaries)

Index(['Unnamed: 0', 'rank', 'discipline', 'yrs.since.phd', 'yrs.service',
       'sex', 'salary'],
      dtype='object')
sex
Female    101002.410256
Male      115090.418994
Name: salary, dtype: float64


In [86]:
salary_by_sex = df.groupby("sex")["salary"].mean()
salary_by_sex

,salary
sex,
Female,101002.410256
Male,115090.418994


Expected results:

| sex | mean salary |
|---|---:|
| Female | 101002.410256 |
| Male | 115090.418994 |

These are **descriptive statistics for the dataset**. They do not establish a causal explanation for the difference.

## E5. Practical Question 2 — Mean salary by academic rank

In [87]:
salary_by_rank =  df.groupby("rank")["salary"].mean()
salary_by_rank

,salary
rank,
AssocProf,93876.437500
AsstProf,80775.985075
Prof,126772.109023


Expected results:

| rank | mean salary |
|---|---:|
| AssocProf | 93876.437500 |
| AsstProf | 80775.985075 |
| Prof | 126772.109023 |

## E6. Practical Question 3 — Mean years of service by sex

In [88]:
service_by_sex = df.groupby("sex")["yrs.service"].mean()
service_by_sex

,yrs.service
sex,
Female,11.564103
Male,18.273743


Expected results:

| sex | mean yrs.service |
|---|---:|
| Female | 11.564103 |
| Male | 18.273743 |

## E7. Multi-key grouping

A list of grouping columns creates more specific subgroups.

For example:

```python
df.groupby(["sex", "rank"])["salary"].mean()
```

This groups records by every combination of `sex` and `rank`.

In [89]:
#Group the rows by discipline and rank and calculate the mean salary
df.groupby(["discipline", "rank"])["salary"].mean()

discipline  rank     
A           AssocProf     83061.115385
            AsstProf      73935.541667
            Prof         119948.274809
B           AssocProf    101276.394737
            AsstProf      84593.906977
            Prof         133393.755556
Name: salary, dtype: float64

## E8. Three-key grouping

Expand the grouping to:

- `sex`
- `rank`
- `discipline`

and calculate the mean of numerical attributes.
The result uses a hierarchical index (MultiIndex).

In [90]:
multi_group = df.groupby(["sex", "rank", "discipline"]).mean(numeric_only=True)

multi_group

Unnamed: 0  yrs.since.phd  yrs.service  \
sex    rank      discipline                                           
Female AssocProf A           128.500000      18.500000    15.500000   
                 B           212.666667      13.500000     8.833333   
       AsstProf  A           191.500000       4.833333     2.500000   
                 B            79.000000       6.600000     2.600000   
       Prof      A           227.750000      26.500000    16.125000   
                 B           151.300000      21.500000    17.900000   
Male   AssocProf A           233.090909      17.727273    13.136364   
                 B           143.468750      13.875000    11.281250   
       AsstProf  A           275.444444       5.944444     2.388889   
                 B           128.210526       4.552632     2.315789   
       Prof      A           256.235772      30.747967    24.983740   
                 B           170.144000      26.552000    21.504000   

                                    salary  
sex    rank      discipline                 
Female AssocProf A            72128.500000  
                 B            99435.666667  
       AsstProf  A            72933.333333  
                 B            84189.800000  
       Prof      A           109631.875000  
                 B           131836.200000  
Male   AssocProf A            85048.863636  
                 B           101621.531250  
       AsstProf  A            74269.611111  
                 B            84647.078947  
       Prof      A           120619.260163  
                 B           133518.360000

Expected mean salary values include:

| sex | rank | discipline | mean salary |
|---|---|---|---:|
| Female | AssocProf | A | 72128.50 |
| Female | AssocProf | B | 99435.67 |
| Female | AsstProf | A | 72933.33 |
| Female | AsstProf | B | 84189.80 |
| Female | Prof | A | 109631.88 |
| Female | Prof | B | 131836.20 |
| Male | AssocProf | A | 85048.86 |
| Male | AssocProf | B | 101621.53 |
| Male | AsstProf | A | 74269.61 |
| Male | AsstProf | B | 84647.08 |
| Male | Prof | A | 120619.26 |
| Male | Prof | B | 133518.36 |

## E9. Reset a MultiIndex result for easier tabular use

This is an explanatory extension useful in later visualization work.

`reset_index()` converts grouping keys back into ordinary columns.

In [91]:
multi_group_table = multi_group.reset_index()
multi_group_table

,sex,rank,discipline,Unnamed: 0,yrs.since.phd,yrs.service,salary
0,Female,AssocProf,A,128.500000,18.500000,15.500000,72128.500000
1,Female,AssocProf,B,212.666667,13.500000,8.833333,99435.666667
2,Female,AsstProf,A,191.500000,4.833333,2.500000,72933.333333
3,Female,AsstProf,B,79.000000,6.600000,2.600000,84189.800000
4,Female,Prof,A,227.750000,26.500000,16.125000,109631.875000
5,Female,Prof,B,151.300000,21.500000,17.900000,131836.200000
6,Male,AssocProf,A,233.090909,17.727273,13.136364,85048.863636
7,Male,AssocProf,B,143.468750,13.875000,11.281250,101621.531250
8,Male,AsstProf,A,275.444444,5.944444,2.388889,74269.611111
9,Male,AsstProf,B,128.210526,4.552632,2.315789,84647.078947


# Part F — Boolean Filtering and Aggregation

## F1. Boolean filtering with multiple conditions

The lecture filters for records satisfying **both** of these conditions:

1. `rank == "Prof"`
2. `yrs.service < 10`

Use the ampersand operator `&` for element-wise AND, and place each comparison in parentheses.

In [92]:
df_prof_10 = df[
    (df["rank"] == "Prof") &
    (df["yrs.service"] < 10)
]

df_prof_10.head()

,Unnamed: 0,rank,discipline,yrs.since.phd,yrs.service,sex,salary
15,16,Prof,B,12,3,Male,117150
25,26,Prof,A,21,8,Male,106294
29,30,Prof,B,12,8,Male,118223
30,31,Prof,B,20,4,Male,132261
32,33,Prof,B,13,9,Male,117256


## F2. Mean salary for professors with fewer than 10 years of service

In [93]:
prof_under_10_mean = df_prof_10["salary"].mean()
prof_under_10_mean

np.float64(125882.27777777778)

Expected result:

```text
125882.27777777778
```

## F3. Count professors with fewer than 10 years of service

The lecture source notes a wording mismatch: an on-screen comment asks for the **total number** of records, while the executed code calculates the **mean salary**.

To answer the count question, use either:

```python
df_prof_10.shape[0]
```

or:

```python
len(df_prof_10)
```

In [94]:
prof_under_10_count = df_prof_10.shape[0]

print("Count:", prof_under_10_count)
print("Count using len():", len(df_prof_10))

Count: 36
Count using len(): 36


Expected count:

```text
36
```

## F4. Professors with more than 10 years of service

Now change the service condition to `> 10`.

In [95]:
df_prof_more_10 = df[
    (df["rank"] == "Prof") &
    (df["yrs.service"] > 10)
]

df_prof_more_10.head()

,Unnamed: 0,rank,discipline,yrs.since.phd,yrs.service,sex,salary
0,1,Prof,B,19,18,Male,139750
1,2,Prof,B,20,16,Male,173200
3,4,Prof,B,45,39,Male,115000
4,5,Prof,B,40,41,Male,141500
6,7,Prof,B,30,23,Male,175000


## F5. Mean salary for professors with more than 10 years of service

In [96]:
prof_more_10_mean = df_prof_more_10["salary"].mean()
prof_more_10_mean

np.float64(127248.65178571429)

Expected result:

```text
127248.65178571429
```

## F6. Count professors with more than 10 years of service

In [97]:
prof_more_10_count = len(df_prof_more_10)
prof_more_10_count

224

Expected count:

```text
224
```

## F7. Compare the two filtered professor groups

The source uses strict conditions `< 10` and `> 10`.

Therefore, professors with exactly **10 years of service** are excluded from both subsets.

In [98]:
comparison = pd.DataFrame({
    "Professor Group": [
        "yrs.service < 10",
        "yrs.service > 10"
    ],
    "Count": [
        prof_under_10_count,
        prof_more_10_count
    ],
    "Mean Salary": [
        prof_under_10_mean,
        prof_more_10_mean
    ]
})

comparison

,Professor Group,Count,Mean Salary
0,yrs.service < 10,36,125882.277778
1,yrs.service > 10,224,127248.651786


### Interpretation

The mean salaries of these filtered subsets are descriptive results from `Salaries.csv`.

A difference between group averages does **not** by itself establish why the difference exists.

## F8. Check professors with exactly 10 years of service

This extension helps explain why the two strict-condition counts do not necessarily sum to the total number of professors.

In [99]:
df_prof_exactly_10 = df[
    (df["rank"] == "Prof") &
    (df["yrs.service"] == 10)
]

print("Professors with exactly 10 years of service:", len(df_prof_exactly_10))

Professors with exactly 10 years of service: 6


# Part H — Consolidated EDA Workflow

The following sequence collects the principal operations demonstrated across the four lecture sections.

In [100]:
# 1. Load data
df = pd.read_csv("Salaries.csv")

# 2. Basic inspection
display(df.head())
print("Shape:", df.shape)
print("Columns:", list(df.columns))
display(df.dtypes)
display(df.values)

# 3. Descriptive statistics
display(df.describe())
print("Mean salary:", df["salary"].mean())
print("Maximum salary:", df["salary"].max())
print("Maximum years of service:", df["yrs.service"].max())

# 4. Column selection
display(df["rank"])
display(df[["yrs.service", "salary"]])
display(df[["yrs.service", "salary", "yrs.since.phd"]].min())

# 5. Categorical exploration
print("Unique rank values:", df["rank"].unique())
display(df["rank"].value_counts())

# 6. Grouped analysis
display(df.groupby("rank").first())
display(df.groupby("rank")["salary"].mean())
display(df.groupby("sex")["salary"].mean())
display(df.groupby("sex")["yrs.service"].mean())
display(df.groupby(["sex", "rank", "discipline"]).mean(numeric_only=True))

# 7. Conditional filtering + aggregation
df_prof_10 = df[
    (df["rank"] == "Prof") &
    (df["yrs.service"] < 10)
]
print("Prof, service < 10 — count:", len(df_prof_10))
print("Prof, service < 10 — mean salary:", df_prof_10["salary"].mean())

df_prof_more_10 = df[
    (df["rank"] == "Prof") &
    (df["yrs.service"] > 10)
]
print("Prof, service > 10 — count:", len(df_prof_more_10))
print("Prof, service > 10 — mean salary:", df_prof_more_10["salary"].mean())

,Unnamed: 0,rank,discipline,yrs.since.phd,yrs.service,sex,salary
0,1,Prof,B,19,18,Male,139750
1,2,Prof,B,20,16,Male,173200
2,3,AsstProf,B,4,3,Male,79750
3,4,Prof,B,45,39,Male,115000
4,5,Prof,B,40,41,Male,141500


Shape: (397, 7)
Columns: ['Unnamed: 0', 'rank', 'discipline', 'yrs.since.phd', 'yrs.service', 'sex', 'salary']


,0
Unnamed: 0,int64
rank,object
discipline,object
yrs.since.phd,int64
yrs.service,int64
sex,object
salary,int64


array([[1, 'Prof', 'B', ..., 18, 'Male', 139750],
       [2, 'Prof', 'B', ..., 16, 'Male', 173200],
       [3, 'AsstProf', 'B', ..., 3, 'Male', 79750],
       ...,
       [395, 'Prof', 'A', ..., 25, 'Male', 101738],
       [396, 'Prof', 'A', ..., 15, 'Male', 95329],
       [397, 'AsstProf', 'A', ..., 4, 'Male', 81035]], dtype=object)

,Unnamed: 0,yrs.since.phd,yrs.service,salary
count,397.000000,397.000000,397.000000,397.000000
mean,199.000000,22.314861,17.614610,113706.458438
std,114.748275,12.887003,13.006024,30289.038695
min,1.000000,1.000000,0.000000,57800.000000
25%,100.000000,12.000000,7.000000,91000.000000
50%,199.000000,21.000000,16.000000,107300.000000
75%,298.000000,32.000000,27.000000,134185.000000
max,397.000000,56.000000,60.000000,231545.000000


Mean salary: 113706.45843828715
Maximum salary: 231545
Maximum years of service: 60


,rank
0,Prof
1,Prof
2,AsstProf
3,Prof
4,Prof
...,...
392,Prof
393,Prof
394,Prof
395,Prof


,yrs.service,salary
0,18,139750
1,16,173200
2,3,79750
3,39,115000
4,41,141500
...,...,...
392,30,103106
393,19,150564
394,25,101738
395,15,95329


,0
yrs.service,0
salary,57800
yrs.since.phd,1


Unique rank values: ['Prof' 'AsstProf' 'AssocProf']


,count
rank,
Prof,266
AsstProf,67
AssocProf,64


,Unnamed: 0,discipline,yrs.since.phd,yrs.service,sex,salary
rank,,,,,,
AssocProf,6,B,6,6,Male,97000
AsstProf,3,B,4,3,Male,79750
Prof,1,B,19,18,Male,139750


,salary
rank,
AssocProf,93876.437500
AsstProf,80775.985075
Prof,126772.109023


,salary
sex,
Female,101002.410256
Male,115090.418994


,yrs.service
sex,
Female,11.564103
Male,18.273743


Unnamed: 0  yrs.since.phd  yrs.service  \
sex    rank      discipline                                           
Female AssocProf A           128.500000      18.500000    15.500000   
                 B           212.666667      13.500000     8.833333   
       AsstProf  A           191.500000       4.833333     2.500000   
                 B            79.000000       6.600000     2.600000   
       Prof      A           227.750000      26.500000    16.125000   
                 B           151.300000      21.500000    17.900000   
Male   AssocProf A           233.090909      17.727273    13.136364   
                 B           143.468750      13.875000    11.281250   
       AsstProf  A           275.444444       5.944444     2.388889   
                 B           128.210526       4.552632     2.315789   
       Prof      A           256.235772      30.747967    24.983740   
                 B           170.144000      26.552000    21.504000   

                                    salary  
sex    rank      discipline                 
Female AssocProf A            72128.500000  
                 B            99435.666667  
       AsstProf  A            72933.333333  
                 B            84189.800000  
       Prof      A           109631.875000  
                 B           131836.200000  
Male   AssocProf A            85048.863636  
                 B           101621.531250  
       AsstProf  A            74269.611111  
                 B            84647.078947  
       Prof      A           120619.260163  
                 B           133518.360000

Prof, service < 10 — count: 36
Prof, service < 10 — mean salary: 125882.27777777778
Prof, service > 10 — count: 224
Prof, service > 10 — mean salary: 127248.65178571429


# Part I — Guided Laboratory Exercises

Complete each exercise in a new code cell.

## Exercise 1 — Dataset integrity
1. Display the first 8 rows.
2. Display the last 5 rows.
3. Print the number of rows.
4. Print the number of columns.
5. Verify that the DataFrame has 397 records.

**Hint:** Use `head()`, `tail()`, and `shape`.

In [105]:
# Exercise 1 — Write your solution below
df.head(8)

,Unnamed: 0,rank,discipline,yrs.since.phd,yrs.service,sex,salary
0,1,Prof,B,19,18,Male,139750
1,2,Prof,B,20,16,Male,173200
2,3,AsstProf,B,4,3,Male,79750
3,4,Prof,B,45,39,Male,115000
4,5,Prof,B,40,41,Male,141500
5,6,AssocProf,B,6,6,Male,97000
6,7,Prof,B,30,23,Male,175000
7,8,Prof,B,45,45,Male,147765


In [107]:
df.tail(5)


,Unnamed: 0,rank,discipline,yrs.since.phd,yrs.service,sex,salary
392,393,Prof,A,33,30,Male,103106
393,394,Prof,A,31,19,Male,150564
394,395,Prof,A,42,25,Male,101738
395,396,Prof,A,25,15,Male,95329
396,397,AsstProf,A,8,4,Male,81035


In [111]:
print("Rows:", df.shape[0])

Rows: 397


In [112]:
print("Columns:", df.shape[1])

Columns: 7


In [114]:
print(df.shape[0])

397


## Exercise 2 — Data types
1. Display the dtype of `yrs.service`.
2. Display the dtype of `rank`.
3. Display all DataFrame dtypes.
4. Explain why arithmetic aggregations can be applied directly to `salary`.

In [115]:
# Exercise 2 — Write your solution below
print(df["yrs.service"].dtype)

int64


In [116]:
print(df["rank"].dtype)


object


In [117]:
print(df.dtypes)

Unnamed: 0        int64
rank             object
discipline       object
yrs.since.phd     int64
yrs.service       int64
sex              object
salary            int64
dtype: object


Explanation

Arithmetic aggregations (like mean, sum, min, max) can be applied directly to salary because it is stored as a numeric data type (typically int64 or float64). Numeric dtypes support mathematical operations, whereas categorical/object types (like rank) do not.

## Exercise 3 — Descriptive statistics
Calculate the following:

1. Minimum salary
2. Maximum salary
3. Mean salary
4. Median salary
5. Standard deviation of salary
6. Minimum years since PhD
7. Maximum years since PhD

In [119]:
# Exercise 3 — Write your solution below
print(df["salary"].min())
print(df["salary"].max())
print(df["salary"].mean())
print(df["salary"].median())
print(df["salary"].std())
print(df["yrs.since.phd"].min())
print(df["yrs.since.phd"].max())

57800
231545
113706.45843828715
107300.0
30289.038694747804
1
56


## Exercise 4 — Column selection
Create:

1. A Series containing only `discipline`.
2. A DataFrame containing only `discipline`.
3. A DataFrame containing `rank`, `sex`, and `salary`.

Then print the Python type of each result.

In [122]:
# Exercise 4 — Write your solution below
discipline_series = df["discipline"]
discipline_df = df[["discipline"]]
selected_df = df[["rank", "sex", "salary"]]

print(type(discipline_series))
print(type(discipline_df))
print(type(selected_df))

<class 'pandas.core.series.Series'>
<class 'pandas.core.frame.DataFrame'>
<class 'pandas.core.frame.DataFrame'>


## Exercise 5 — Explore categorical variables
For each of the following columns:

- `rank`
- `discipline`
- `sex`

1. Display the unique categories.
2. Display their frequencies with `value_counts()`.

In [127]:
# Exercise 5 — Write your solution below
print(df["rank"].unique())
print(df["rank"].value_counts())

['Prof' 'AsstProf' 'AssocProf']
rank
Prof         266
AsstProf      67
AssocProf     64
Name: count, dtype: int64


In [126]:
print(df["discipline"].unique())
print(df["discipline"].value_counts())

['B' 'A']
discipline
B    216
A    181
Name: count, dtype: int64


In [128]:
print(df["sex"].unique())
print(df["sex"].value_counts())

['Male' 'Female']
sex
Male      358
Female     39
Name: count, dtype: int64


## Exercise 6 — GroupBy practice
Calculate:

1. Mean salary by `discipline`.
2. Mean salary by `sex`.
3. Mean years since PhD by `rank`.
4. Mean years of service by `rank`.

In [130]:
# Exercise 6 — Write your solution below
print(df.groupby("discipline")["salary"].mean())

discipline
A    108548.430939
B    118028.694444
Name: salary, dtype: float64


In [131]:
print(df.groupby("sex")["salary"].mean())

sex
Female    101002.410256
Male      115090.418994
Name: salary, dtype: float64


In [132]:
print(df.groupby("rank")["yrs.since.phd"].mean())

rank
AssocProf    15.453125
AsstProf      5.104478
Prof         28.300752
Name: yrs.since.phd, dtype: float64


In [133]:
print(df.groupby("rank")["yrs.service"].mean())

rank
AssocProf    11.953125
AsstProf      2.373134
Prof         22.815789
Name: yrs.service, dtype: float64


## Exercise 7 — Multi-key GroupBy
Calculate mean salary for each combination of:

1. `rank` and `discipline`
2. `sex` and `discipline`
3. `sex`, `rank`, and `discipline`

Then reset the index for the three-key result.

In [134]:
# Exercise 7 — Write your solution below
print(df.groupby(["rank", "discipline"])["salary"].mean())

rank       discipline
AssocProf  A              83061.115385
           B             101276.394737
AsstProf   A              73935.541667
           B              84593.906977
Prof       A             119948.274809
           B             133393.755556
Name: salary, dtype: float64


In [135]:
print(df.groupby(["sex", "discipline"])["salary"].mean())

sex     discipline
Female  A              89064.944444
        B             111234.523810
Male    A             110699.981595
        B             118760.374359
Name: salary, dtype: float64


In [136]:
multi = df.groupby(["sex", "rank", "discipline"])["salary"].mean()
print(multi.reset_index())

       sex       rank discipline         salary
0   Female  AssocProf          A   72128.500000
1   Female  AssocProf          B   99435.666667
2   Female   AsstProf          A   72933.333333
3   Female   AsstProf          B   84189.800000
4   Female       Prof          A  109631.875000
5   Female       Prof          B  131836.200000
6     Male  AssocProf          A   85048.863636
7     Male  AssocProf          B  101621.531250
8     Male   AsstProf          A   74269.611111
9     Male   AsstProf          B   84647.078947
10    Male       Prof          A  120619.260163
11    Male       Prof          B  133518.360000


## Exercise 8 — Boolean filtering
Create a subset containing:

- only `AssocProf` records
- with more than 10 years of service

Then calculate:

1. number of matching records
2. mean salary
3. minimum salary
4. maximum salary

In [139]:
# Exercise 8 — Write your solution below
subset = df[(df["rank"] == "AssocProf") & (df["yrs.service"] > 10)]
print(len(subset))
print(subset["salary"].mean())
print(subset["salary"].min())
print(subset["salary"].max())

18
91041.27777777778
62884
126431


## Exercise 9 — Combined conditions
Create a subset for:

- `discipline == "B"`
- `sex == "Female"`

Then calculate the average:

1. salary
2. years of service
3. years since PhD

In [140]:
# Exercise 9 — Write your solution below
subset = df[(df["discipline"] == "B") & (df["sex"] == "Female")]

print(subset["salary"].mean())
print(subset["yrs.service"].mean())
print(subset["yrs.since.phd"].mean())

111234.52380952382
11.666666666666666
15.666666666666666


## Exercise 10 — Validate a categorical count

Verify programmatically that:

```python
df["rank"].value_counts().sum()
```

is equal to the number of DataFrame rows.

Your code should print either `True` or `False`.

In [141]:
# Exercise 10 — Write your solution below
print(df["rank"].value_counts().sum() == df.shape[0])

True


# Laboratory Summary

This laboratory implemented the complete EDA workflow developed across the Week 4 pandas lecture material:

- **Load** data with `pd.read_csv()`.
- **Inspect** structure with `head()`, `shape`, `columns`, `dtypes`, and `values`.
- **Summarize** numeric variables with `describe()`, `mean()`, `min()`, `max()`, `median()`, and `std()`.
- **Select** one or more columns with bracket notation.
- **Explore categories** with `unique()` and `value_counts()`.
- **Group** records with `groupby()` and calculate aggregate statistics.
- **Use multiple grouping keys** to create more detailed subgroup comparisons.
- **Filter** records with Boolean conditions.
- **Aggregate filtered subsets** to answer focused analytical questions.
- **Recognize errors** caused by import typos, wrong column names, and incorrect grouping syntax.
- **Interpret results carefully** as descriptions of the available dataset.

The central EDA principle is to understand the DataFrame's structure, types, categories, missingness, and numerical summaries before relying on later visualizations or interpretations.